cfpb/consumerfinance.gov
# Reports as webpages: initial test

This is an initial feature set for testing the Reports as Webpages process in our Build environment. Collaboration with @wpears.

## Additions

- New django app called `research_reports` for the reports models and code
- New page type, `ResearchReportPage`, with associated child models, e.g. `ReportSection`, `ReportAuthor`
- Jinja templates for displaying `ResearchReportPage`s
- Front-end code for creating and displaying a table of contents in the left-hand sidebar
- New management command for fetching a Word report, converting its contents to HTML, and putting its report data into a `ResearchReportPage` instance
- New management command for calling the Jenkins API to kick off processing a report


## Todos

Reports as Webpages code is not a finished feature. Here's an incomplete list of additional things we'll be adding in future PRs:

- Code to convert all remaining Report content from Pandoc-produced HTML to `ResearchReportPage` data fields, including:
    - Content sections, subsections, and third-level sections
    - Appendices
    - Figures
    - Data tables
- Models and front-end handling of a third level of report content
- Python tests
- Error handling, including scenarios where: 
    - Jenkins is down (when on a server where connection to Jenkins is enabled)
    - Pandoc isn't installed (when running in a location where local processing is enabled)
    - The provided Word report isn't parseable as a report
- Content protections to ensure original Word reports aren't discoverable by the public




## How to test this PR

1. Install Pandoc in your local environment: `brew install pandoc`
1. Get a sample report: [GHE]/CFPB/super-regular/wiki/Reports-as-Webpages:-sample-reports-for-testing
2. Get a recent database dump. Run migrations on this branch. `./frontend.sh`.
3. Run the branch, log in to the Wagtail Admin
4. Create a new Research Report Page
5. On the "Report Upload" tab, enter a Title manually. Using the "Report file" file chooser, select your sample report. Check the "Process report" checkbox.
6. Click the "Save draft" button. Within 10 seconds, the page banner should indicate that it was updated.
7. Visit the "Report Content" tab, see that the following content from the sample report was inserted into the page fields:
    - Report type, title, subtitle, authors, and footnotes.
8. Press the "Preview" button, see the rendered version of the report, including interactive sidebar table of contents


## Screenshots
![Screen Shot 2020-08-03 at 1 20 19 PM](https://user-images.githubusercontent.com/4295388/89209351-1c96bd00-d58c-11ea-9ba0-be6efe534625.png)
![Screen Shot 2020-08-03 at 1 20 58 PM](https://user-images.githubusercontent.com/4295388/89209389-2fa98d00-d58c-11ea-8ec8-53f7f0386262.png)


## Checklist
